src/algebra/invertible.lean

```diff
@@ -0,0 +1,184 @@
+/-
+Copyright (c) 2020 Anne Baanen. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Anne Baanen
+
+A typeclass for the two-sided multiplicative inverse.
+-/
+
+import tactic
+
+/-!
+# Invertible elements
+
+This file defines a typeclass `invertible a` for elements `a` with a
+multiplicative inverse.
+
+The intent of the typeclass is to provide a way to write e.g. `⅟2` in a ring
+like `ℤ[1/2]` where some inverses exist but there is no general `⁻¹` operator;
+or to specify that a field has characteristic `≠ 2`.
+It is the `Type`-valued analogue to the `Prop`-valued `is_unit`.
+
+This file also includes some instances of `invertible` for specific numbers in
+characteristic zero. Some more cases are given as a `def`, to be included only
+when needed. To construct instances for concrete numbers,
+`invertible_of_nonzero` is a useful definition.
+
+## Notation
+
+ * `⅟a` is `invertible.inv_of a`, the inverse of `a`
+
+## Implementation notes
+
+The `invertible` class lives in `Type`, not `Prop`, to make computation easier.
+If multiplication is associative, `invertible` is a subsingleton anyway.
+
+The `simp` normal form tries to normalize `⅟a` to `a ⁻¹`. Otherwise, it pushes
+`⅟` inside the expression as much as possible.
+
+## Tags
+
+invertible, inverse element, inv_of, a half, one half, a third, one third, ½, ⅓
+
+-/
+
+universes u
+
+variables {α : Type u}
+
+/-- `invertible a` gives a two-sided multiplicative inverse of `a`. -/
+class invertible [has_mul α] [has_one α] (a : α) : Type u :=
+(inv_of : α) (inv_of_mul_self : inv_of * a = 1) (mul_inv_of_self : a * inv_of = 1)
+
+-- This notation has the same precedence as `has_inv.inv`.
+notation `⅟`:1034 := invertible.inv_of
+
+@[simp]
+lemma inv_of_mul_self [has_mul α] [has_one α] (a : α) [invertible a] : ⅟a * a = 1 :=
+invertible.inv_of_mul_self
+
+@[simp]
+lemma mul_inv_of_self [has_mul α] [has_one α] (a : α) [invertible a] : a * ⅟a = 1 :=
+invertible.mul_inv_of_self
+
+@[simp]
+lemma mul_inv_of_mul_self_cancel [monoid α] (a b : α) [invertible b] : a * ⅟b * b = a :=
+by simp [mul_assoc]
+
+@[simp]
+lemma mul_mul_inv_of_self_cancel [monoid α] (a b : α) [invertible b] : a * b * ⅟b = a :=
+by simp [mul_assoc]
+
+lemma left_inv_eq_right_inv [monoid α] {a b c : α} (hba : b * a = 1) (hac : a * c = 1) : b = c :=
```

Comment:
Could you move this to a more proper place?

Reply:
Is `algebra/group/basic.lean` a suitable location?

Reply:
Yes, this would be perfect